central/cloudsources/manager/manager.go

```diff
@@ -138,18 +150,64 @@ func (m *managerImpl) getDiscoveredClustersFromCloudSources() []*discoveredclust
 		log.Errorw("Received errors during fetching assets from cloud sources. The result might be incomplete",
 			logging.Err(clientErrors))
 	}
-	log.Debugf("Got the following discovered clusters from Cloud Source integrations: %+v", discoveredClusters)
+	debugPrintDiscoveredClusters(discoveredClusters)
 	return discoveredClusters
 }
 
 func (m *managerImpl) reconcileDiscoveredClusters(clusters []*discoveredclusters.DiscoveredCluster) {
-	// TODO: Add matching of discovered clusters with secured clusters.
+	m.matchDiscoveredClusters(clusters)
+
+	debugPrintDiscoveredClusters(clusters)
 
 	if err := m.discoveredClustersDataStore.UpsertDiscoveredClusters(discoveredClusterCtx, clusters...); err != nil {
 		log.Errorw("Received errors during upserting discovered clusters.", logging.Err(err))
 	}
 }
 
+func (m *managerImpl) matchDiscoveredClusters(clusters []*discoveredclusters.DiscoveredCluster) {
+	// A list of hashes of currently secured cluster. A secured cluster hash consist of:
+	//  The secured cluster ID, name, and type.
+	securedClusters := set.NewStringSet()
+
+	unspecifiedClusterTypes := set.NewStringSet()
+
+	if err := m.clusterDataStore.WalkClusters(clustersCtx, func(obj *storage.Cluster) error {
+		providerMetadata := obj.GetStatus().GetProviderMetadata()
+
+		// Explicitly ignore clusters which do not have provider metadata associated with them.
+		if providerMetadata == nil {
```

Comment:
This will be a source of false positives because of old sensors. But I guess we can't do anything about it otherwise than tell customers to upgrade their sensors to get correct matching.